generated/generated.go

```diff
@@ -15,67 +14,57 @@
 //
-// 	The text must appear as the first line of a properly formatted Go // comment,
-// 	and that comment must appear before but not be attached to the package clause
-// 	and before any /* */ comment. This is similar to the rules for build tags.
-//
-// 	The comment line must match the case-sensitive regular expression (in Go syntax):
+// 	Generated files are marked by a line of text that matches
+// 	the regular expression, in Go syntax:
 //
 // 		`^// Code generated .* DO NOT EDIT\.$`
 //
 // 	The .* means the tool can put whatever folderol it wants in there,
 // 	but the comment must be a single line and must start with `Code generated`
 // 	and end with `DO NOT EDIT.`, with a period.
 //
-// If the source couldn't be read, the error indicates the specific
-// failure. If the source was read but syntax errors were found,
-// the result is estimated on a best effort basis from a partial AST.
+// 	The text may appear anywhere in the file.
 //
-// TODO: Decide on best policy of what to do in case of syntax errors
-// being encountered during parsing.
+// If the source couldn't be read, the error indicates the specific
+// failure.
 func ParseFile(filename string) (hasGeneratedComment bool, err error) {
-	fset := token.NewFileSet()
-	f, err := parser.ParseFile(fset, filename, nil, parser.PackageClauseOnly|parser.ParseComments)
-	if f == nil { // Can only happen when err != nil.
+	f, err := os.Open(filename)
+	if err != nil {
 		return false, err
 	}
-Outer:
-	for _, cg := range f.Comments {
-		if cg == f.Doc {
-			// If we've reached the package comment, don't look any further,
-			// because the generated comment must be before that.
-			break
+	defer f.Close()
+	br := bufio.NewReader(f)
+	for {
+		s, err := br.ReadBytes('\n')
+		if err == io.EOF {
+			return containsComment(s), nil
+		} else if err != nil {
+			return false, err
 		}
-		// Check if this comment group is a match.
-		// The text must appear as the first line of a properly formatted line comment (//-style).
-		if len(cg.List[0].Text) >= smallestMatchingComment &&
-			strings.HasPrefix(cg.List[0].Text, "// Code generated ") &&
-			strings.HasSuffix(cg.List[0].Text, " DO NOT EDIT.") &&
-			fset.Position(cg.List[0].Pos()).Column == 1 {
-
+		s = s[:len(s)-1] // Trim newline.
+		if containsComment(s) {
 			return true, nil
 		}
-		// Ensure none of the comments in this comment group are general comments (/*-style).
-		for _, c := range cg.List {
-			if strings.HasPrefix(c.Text, "/*") {
-				// If we've reached a general comment (/*-style), don't look any further,
-				// because the generated comment must be before that.
-				break Outer
-			}
-		}
 	}
-	return false, nil
+}
+
+// containsComment reports whether a line of Go source code s (without newline character)
+// contains the generated comment.
+func containsComment(s []byte) bool {
+	return len(s) >= smallestMatchingComment &&
+		bytes.HasPrefix(s, []byte("// Code generated ")) &&
```

Comment:
I guess it is getting inlined:

```
bradfitz@gdev:~/src/bytesprefix$ GOROOT=$HOME/go1.7 go tool compile -m bench_test.go
bench_test.go:18: inlining call to ReportAllocs
bench_test.go:31: inlining call to ReportAllocs
bench_test.go:13: matchVar line does not escape
bench_test.go:17: BenchmarkHasPrefixVar b does not escape
bench_test.go:26: matchInline line does not escape
bench_test.go:27: matchInline ([]byte)("nope") does not escape
bench_test.go:30: BenchmarkHasPrefixInline b does not escape
bradfitz@gdev:~/src/bytesprefix$ go tool compile -m bench_test.go
bench_test.go:18:16: inlining call to (*testing.B).ReportAllocs
bench_test.go:31:16: inlining call to (*testing.B).ReportAllocs
bench_test.go:13:28: matchVar line does not escape
bench_test.go:17:31: BenchmarkHasPrefixVar b does not escape
bench_test.go:26:31: matchInline line does not escape
bench_test.go:27:37: matchInline ([]byte)("nope") does not escape
bench_test.go:30:34: BenchmarkHasPrefixInline b does not escape
<autogenerated>:1:0: leaking param: .this
<autogenerated>:1:0: leaking param: .this
```

But it's still faster with the var:

```
$ go test -v -bench=. -benchtime=2s
goos: linux
goarch: amd64
pkg: bytesprefix
BenchmarkHasPrefixVar-4         300000000               10.6 ns/op             0 B/op          0 allocs/op
BenchmarkHasPrefixInline-4      100000000               21.5 ns/op             0 B/op          0 allocs/op
```

Compare the variable version:

```
"".matchVar t=1 size=125 args=0x20 locals=0x40
        0x0000 00000 (bench_test.go:13) TEXT    "".matchVar(SB), $64-32
        0x0000 00000 (bench_test.go:13) MOVQ    (TLS), CX
        0x0009 00009 (bench_test.go:13) CMPQ    SP, 16(CX)
        0x000d 00013 (bench_test.go:13) JLS     118
        0x000f 00015 (bench_test.go:13) SUBQ    $64, SP
        0x0013 00019 (bench_test.go:13) MOVQ    BP, 56(SP)
        0x0018 00024 (bench_test.go:13) LEAQ    56(SP), BP
        0x001d 00029 (bench_test.go:13) FUNCDATA        $0, gclocals·4032f753396f2012ad1784f398b170f4(SB)
        0x001d 00029 (bench_test.go:13) FUNCDATA        $1, gclocals·69c1753bd5f81501d95132d08af04464(SB)
        0x001d 00029 (bench_test.go:14) MOVQ    "".line+72(SP), AX
        0x0022 00034 (bench_test.go:14) MOVQ    AX, (SP)
        0x0026 00038 (bench_test.go:14) MOVQ    "".line+80(SP), AX
        0x002b 00043 (bench_test.go:14) MOVQ    AX, 8(SP)
        0x0030 00048 (bench_test.go:14) MOVQ    "".line+88(SP), AX
        0x0035 00053 (bench_test.go:14) MOVQ    AX, 16(SP)
        0x003a 00058 (bench_test.go:14) MOVQ    "".prefix(SB), AX
        0x0041 00065 (bench_test.go:14) MOVQ    "".prefix+8(SB), CX
        0x0048 00072 (bench_test.go:14) MOVQ    "".prefix+16(SB), DX
        0x004f 00079 (bench_test.go:14) MOVQ    AX, 24(SP)
        0x0054 00084 (bench_test.go:14) MOVQ    CX, 32(SP)
        0x0059 00089 (bench_test.go:14) MOVQ    DX, 40(SP)
        0x005e 00094 (bench_test.go:14) PCDATA  $0, $1
        0x005e 00094 (bench_test.go:14) CALL    bytes.HasPrefix(SB)
        0x0063 00099 (bench_test.go:14) MOVBLZX 48(SP), AX
        0x0068 00104 (bench_test.go:14) MOVB    AL, "".~r1+96(SP)
        0x006c 00108 (bench_test.go:14) MOVQ    56(SP), BP
        0x0071 00113 (bench_test.go:14) ADDQ    $64, SP
        0x0075 00117 (bench_test.go:14) RET
        0x0076 00118 (bench_test.go:14) NOP
        0x0076 00118 (bench_test.go:13) PCDATA  $0, $-1
        0x0076 00118 (bench_test.go:13) CALL    runtime.morestack_noctxt(SB)
        0x007b 00123 (bench_test.go:13) JMP     0
```

With the copying inline version:

```
"".matchInline t=1 size=164 args=0x20 locals=0x60
        0x0000 00000 (bench_test.go:26) TEXT    "".matchInline(SB), $96-32
        0x0000 00000 (bench_test.go:26) MOVQ    (TLS), CX
        0x0009 00009 (bench_test.go:26) CMPQ    SP, 16(CX)
        0x000d 00013 (bench_test.go:26) JLS     154
        0x0013 00019 (bench_test.go:26) SUBQ    $96, SP
        0x0017 00023 (bench_test.go:26) MOVQ    BP, 88(SP)
        0x001c 00028 (bench_test.go:26) LEAQ    88(SP), BP
        0x0021 00033 (bench_test.go:26) FUNCDATA        $0, gclocals·4032f753396f2012ad1784f398b170f4(SB)
        0x0021 00033 (bench_test.go:26) FUNCDATA        $1, gclocals·69c1753bd5f81501d95132d08af04464(SB)
        0x0021 00033 (bench_test.go:27) LEAQ    ""..autotmp_6+56(SP), AX
        0x0026 00038 (bench_test.go:27) MOVQ    AX, (SP)
        0x002a 00042 (bench_test.go:27) LEAQ    go.string."nope"(SB), AX
        0x0031 00049 (bench_test.go:27) MOVQ    AX, 8(SP)
        0x0036 00054 (bench_test.go:27) MOVQ    $4, 16(SP)
        0x003f 00063 (bench_test.go:27) PCDATA  $0, $0
        0x003f 00063 (bench_test.go:27) CALL    runtime.stringtoslicebyte(SB)
        0x0044 00068 (bench_test.go:27) MOVQ    24(SP), AX
        0x0049 00073 (bench_test.go:27) MOVQ    32(SP), CX
        0x004e 00078 (bench_test.go:27) MOVQ    40(SP), DX
        0x0053 00083 (bench_test.go:27) MOVQ    "".line+104(SP), BX
        0x0058 00088 (bench_test.go:27) MOVQ    BX, (SP)
        0x005c 00092 (bench_test.go:27) MOVQ    "".line+112(SP), BX
        0x0061 00097 (bench_test.go:27) MOVQ    BX, 8(SP)
        0x0066 00102 (bench_test.go:27) MOVQ    "".line+120(SP), BX
        0x006b 00107 (bench_test.go:27) MOVQ    BX, 16(SP)
        0x0070 00112 (bench_test.go:27) MOVQ    AX, 24(SP)
        0x0075 00117 (bench_test.go:27) MOVQ    CX, 32(SP)
        0x007a 00122 (bench_test.go:27) MOVQ    DX, 40(SP)
        0x007f 00127 (bench_test.go:27) PCDATA  $0, $1
        0x007f 00127 (bench_test.go:27) CALL    bytes.HasPrefix(SB)
        0x0084 00132 (bench_test.go:27) MOVBLZX 48(SP), AX
        0x0089 00137 (bench_test.go:27) MOVB    AL, "".~r1+128(SP)
        0x0090 00144 (bench_test.go:27) MOVQ    88(SP), BP
        0x0095 00149 (bench_test.go:27) ADDQ    $96, SP
        0x0099 00153 (bench_test.go:27) RET
        0x009a 00154 (bench_test.go:27) NOP
        0x009a 00154 (bench_test.go:26) PCDATA  $0, $-1
        0x009a 00154 (bench_test.go:26) CALL    runtime.morestack_noctxt(SB)
        0x009f 00159 (bench_test.go:26) JMP     0
```